File: 0087-scramble-string/0087-scramble-string.js
/**
 * @param {string} s1
 * @param {string} s2
 * @return {boolean}
 */
var isScramble = function(s1, s2) {
    const dp = {}
    return helper(s1, s2)

    function helper(s1, s2) {
        const key = `${s1} ${s2}`
        if (key in dp) {
            return dp[key]
        }
        if (s1.length === 1) {
            return s1 === s2
        }
        if (s1 === s2) {
            return true
        }
        const n = s1.length
        for (let i=1; i < n; i++) {
            if ((helper(s1.substring(0, i), s2.substring(0, i)) && helper(s1.substring(i), s2.substring(i))) || 
                (helper(s1.substring(0, i), s2.substring(n-i)) && helper(s1.substring(i), s2.substring(0, n-i)))) {
                return true
            }
        }
        return dp[key] = false
    }
};

File: 0087-scramble-string/0087-scramble-string.py
class Solution:
    def isScramble(self, s1: str, s2: str) -> bool:
        
        def helper(s1: str, s2: str) -> bool:
            key = (s1, s2)
            
            if key in dp:
                return dp[key]
            
            if sorted(s1) != sorted(s2):
                dp[key] = False
                return dp[key]
            
            if s1 == s2:
                dp[key] = True
                return dp[key]

            for i in range(1, len(s1)):
                if (helper(s1[:i:], s2[:i:]) and helper(s1[i::], s2[i::])) or (helper(s1[i::], s2[:-i:]) and helper(s1[:i:], s2[-i::])):
                    dp[key] = True
                    return dp[key]
                
            dp[key] = False
            return dp[key]
        
        dp = {}
        return helper(s1, s2)